tests/integration/test_task.py

```diff
@@ -80,3 +83,245 @@ def _run(self, input_files):
+        fil.write(diagnostic_text)
+
+    task = DiagnosticTask(
+        name='task0',
+        ancestors=None,
+        script=diag_script.as_posix(),
+        settings=diag_settings,
+        output_dir=diag_output_dir.as_posix(),
+    )
+
+    return {task}
+
+
+def test_py_diagnostic_run_sequential_task_fails(monkeypatch, tmp_path):
+    """Run DiagnosticTask sequentially with bad Python script."""
+    diagnostic_text = "import os\n\nprint(cow)"
+    tasks = _get_diagnostic_tasks(tmp_path, diagnostic_text, 'py')
+
+    def _run(self, input_filesi=[]):
+        print(f'running task {self.name}')
+
+    monkeypatch.setattr(BaseTask, '_run', _run)
+
+    with pytest.raises(DiagnosticError) as err_mssg:
+        _run_tasks_sequential(tasks)
+    exp_mssg = "diag_cow.py failed with return code 1"
+    assert exp_mssg in str(err_mssg.value)
+
+
+def test_py_diagnostic_run_parallel_task_fails(monkeypatch, tmp_path):
+    """Run DiagnosticTask parallel with bad Python script."""
+    diagnostic_text = "import os\n\nprint(cow)"
+    tasks = _get_diagnostic_tasks(tmp_path, diagnostic_text, 'py')
+
+    def _run(self, input_filesi=[]):
+        print(f'running task {self.name}')
+
+    monkeypatch.setattr(BaseTask, '_run', _run)
+
+    with pytest.raises(DiagnosticError) as err_mssg:
+        _run_tasks_parallel(tasks, 2)
+    exp_mssg = "diag_cow.py failed with return code 1"
+    assert exp_mssg in str(err_mssg.value)
+
+
+def test_py_diagnostic_run_parallel_task(monkeypatch, tmp_path):
+    """Run DiagnosticTask in parallel with OK Python script."""
+    diagnostic_text = "import os\n\nprint('cow')"
+    tasks = _get_diagnostic_tasks(tmp_path, diagnostic_text, 'py')
+
+    def _run(self, input_filesi=[]):
+        print(f'running task {self.name}')
+
+    monkeypatch.setattr(BaseTask, '_run', _run)
+
+    _run_tasks_parallel(tasks, 2)
+
+
+def test_ncl_diagnostic_run_parallel_task_fails(monkeypatch, tmp_path):
+    """Run DiagnosticTask in parallel with OK NCL script."""
+    diagnostic_text = "cows on the [river]"
+
+    def _run(self, input_filesi=[]):
+        print(f'running task {self.name}')
+
+    if shutil.which('ncl') is not None:
+        tasks = _get_diagnostic_tasks(tmp_path, diagnostic_text, 'ncl')
+
+        monkeypatch.setattr(BaseTask, '_run', _run)
+        with pytest.raises(DiagnosticError) as err_mssg:
+            _run_tasks_parallel(tasks, 2)
+        exp_mssg_1 = "An error occurred during execution of NCL script"
+        exp_mssg_2 = "diag_cow.ncl"
+        assert exp_mssg_1 in str(err_mssg.value)
+        assert exp_mssg_2 in str(err_mssg.value)
+    else:
+        with pytest.raises(DiagnosticError) as err_mssg:
+            tasks = _get_diagnostic_tasks(tmp_path, diagnostic_text, 'ncl')
+        exp_mssg_1 = "Cannot execute script "
+        exp_mssg_2 = "program 'ncl' not installed."
+        assert exp_mssg_1 in str(err_mssg.value)
+        assert exp_mssg_2 in str(err_mssg.value)
```

Comment:
This test would test a smaller amount of code and would be more logically placed if you added it below this test:
https://github.com/ESMValGroup/ESMValCore/blob/54180f7aee4fbad0f6b44cf74963a9d956860c0c/tests/unit/test_task.py#L59-L61

Reply:
I dont get it - we are testing for the existence of ncl and R via `shutil`

Reply:
Sorry for not being clear :blush: What I mean is that the test I referred to above tests just the function `DiagnosticTask._initialize_cmd`, which raises the exception. However, there is no test yet for the cases where an exception is raised, only for the normal case. So from a code organization point of view it would be better to add this test there, because then it's easy to see that both normal run and exception are tested.

Reply:
I still don't get it - you can't really use `parametrize` since the tests differ slightly in terms of the actual diagnostic to be executed - you want to check first for the existence of the executable **and** the interpreter correctly installed? Meh, think this is fine, man

Reply:
This test should be taken apart into two separate tests, the bit if the executable is present can stay here, the bit if the executable is not present should be moved to tests/unit/test_task.py

Reply:
yes makes sense, I thought I can get away but no chance :racehorse:

Reply:
dang! Just realized you were all this time pointing to an existing code in the unit tests! sorry for making you chase around dude, yes, now I completely understand the call